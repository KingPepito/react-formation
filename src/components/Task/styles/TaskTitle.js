import styled from 'styled-components'

const TaskTitle = styled.span`
  width: 650px;
`

export default TaskTitle