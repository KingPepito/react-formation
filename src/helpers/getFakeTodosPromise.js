import {getGUID} from "./index";

export const getFakeTodosPromise = () => new Promise((resolve, reject) => {
  setTimeout(() => {
    // 1 in 5 chance to get a fake error
    const isError = Math.random() < 0.2
    if(isError) {
      reject('An error occured when loading your todo list. Please try again')
      return
    }
    resolve({
      data: [
        {
          title: 'Metro',
          completed: false,
          id: getGUID()
        },
        {
          title: 'Boulot',
          completed: false,
          id: getGUID()
        },
        {
          title: 'Dodo',
          completed: false,
          id: getGUID()
        }
      ]
    })
  }, 1200)
})