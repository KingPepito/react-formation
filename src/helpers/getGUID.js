const nodeCrypto = require('crypto');
// Providing crypto for jest tests, this is a quick fix,
// we might use something else than global.crypto to generate random values
if(typeof crypto === 'undefined') global.crypto = {
  getRandomValues: function(buffer) { return nodeCrypto.randomFillSync(buffer);}
};

function uuidv4() {
  return ([1e7]+-1e3+-4e3+-8e3+-1e11).replace(/[018]/g, c =>
    (c ^ crypto.getRandomValues(new Uint8Array(1))[0] & 15 >> c / 4).toString(16)
  );
}

export default uuidv4