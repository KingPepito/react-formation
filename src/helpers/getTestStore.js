import {applyMiddleware, createStore} from 'redux'
import rootReducer from '../redux/reducers/index'
import {middlewares} from '../redux/configureStore'
// This helper provide an instance of a similar store used in the app.
export const getTestStore = initialState => {
  const createStoreWithMiddleware = applyMiddleware(...middlewares)(createStore)
  return createStoreWithMiddleware(rootReducer, initialState)
}