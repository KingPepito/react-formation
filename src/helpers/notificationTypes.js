export const DEFAULT = 'DEFAULT'
export const ERROR = 'ERROR'
export const VALIDATION = 'VALIDATION'
export const WARNING = 'WARNING'

// Todo: integrate it in the theme?
export const NotificationTypes = {
  default: {
    type: DEFAULT,
    color: '#7dd0f2'
  },
  error: {
    type: ERROR,
    color: '#d2816f'
  },
  validation: {
    type: VALIDATION,
    color: '#87d289'
  },
  warning: {
    type: WARNING,
    color: '#d2bb64'
  },
}