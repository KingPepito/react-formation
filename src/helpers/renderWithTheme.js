import React from 'react';
import renderer from 'react-test-renderer';
import theme from '../theme';
import {ThemeProvider} from 'styled-components';

export const wrapWithThemeProvider = component =>
  <ThemeProvider theme={theme}>
    {component}
  </ThemeProvider>

// Helpers creating a component to be tested with styled component Theme injected
export function renderWithTheme(component) {
  return renderer.create(
    <ThemeProvider theme={theme}>
      {component}
    </ThemeProvider>
  );
};