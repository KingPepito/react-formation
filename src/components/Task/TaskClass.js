import React, {Component} from "react";
import PropTypes from "prop-types";
import TaskContainer from "./styles/TaskContainer";
import TaskTitle from "./styles/TaskTitle";
// This is a class equivalent of Task, as Enzyme does not support interaction with hooks features as "useState".
// For the purpose of our Enzyme test, we would like to store "complete" in the state
// instead of being a prop from the store and passed as a prop.
class TaskClass extends Component {

  state = {
    completed: false
  }

  removeSelf = () => {
    const {id, remove} = this.props
    remove(id)
  }

  completeSelf = () => {
    const {completed} = this.state
    this.setState({completed: !completed})
  }

  render() {
    const {title} = this.props
    const {completed} = this.state

    return <TaskContainer completed={completed}>
      <TaskTitle onClick={this.completeSelf}>{title}</TaskTitle>
      <button onClick={this.removeSelf} data-test='remove-button-task-test'>X</button>
    </TaskContainer>
  }
}

TaskClass.propTypes = {
  complete: PropTypes.func,
  completed: PropTypes.bool,
  id: PropTypes.string,
  title: PropTypes.string,
  remove: PropTypes.func,
}

export default TaskClass