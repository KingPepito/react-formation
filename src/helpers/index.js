import getGUID from './getGUID'

export {
  getGUID
}