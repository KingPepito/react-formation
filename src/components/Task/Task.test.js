import React from 'react';
import {renderWithTheme} from "../../helpers/renderWithTheme";
import {shallow} from "enzyme";
import TaskTitle from "./styles/TaskTitle";
// Here we using a class version of Task. It is because Enzyme does not support manipulation of hooks yet.
// It might be a serious problem to consider when choosing the utilities lib for your React app.
import TaskClass from "./TaskClass";


// We simply testing that our component Task is rendering here, snapshot testing might also include more features...
// Each time we run a snapshot test, it will create a snapshot files inside __snapshots__.
// Snapshot test will notify if the HTML produced by our component has been changes since the last time we run the test.
// If it is wanted that our changes will modify the HTML output, we shall update the tests by using 'yarn test -u'.
describe("Task snapshot test.", () => {
  // "describe" is use to creates a block that groups together several related tests. It is not mandatory but can make the test more readable.
  // "test" is used to assert that the function passed as argument is executed correctly, first arg is describing the test.
  test('Simply test if the component is rendering', () => {
    const component = renderWithTheme(setupTask({title:"Hey I'm a test task!"}));
    let tree = component.toJSON();
    // "expect" like "describe" is a function related to Jest, it might looks unintuitive but no need to import it here.
    // Jest will make sure when running that this kind of functions are imported.
    expect(tree).toMatchSnapshot();
  });
})
// "shallow" is provided by Enzyme, it renders only the single component, not including its children.
// This is useful to isolate the component for pure unit testing.
const setupTask = (props = {}) => shallow(<TaskClass {...props}/>)

describe("Task DOM testing with Enzyme.", () => {
  let componentToTest
  const title = "Hey I'm a second test task!"
  // "beforeEach" is a function that Jest will run before each test of the current block.
  beforeEach(() => {
    // Initialize the Task component with some props.
    componentToTest = setupTask({title})
  })
  // "it" is a alias for "test" function, it allows us to read the test call as "it should render a remove button".
  // Here we assume that the remove button is a really important Task component's feature and its presence need to be test.
  it('should render a remove button', () => {
    // This will prove that our Task component contains a Button.
    expect(componentToTest.find('button'))
    // But we might want to be more specific by using a custom attribute to identify the node we are looking for.
    expect(componentToTest.find("[data-test='remove-button-task-test']"))
  })
  it('should render a title with the expected text', () => {
    // We might find an imported sub-component like this:
    const titleCompo = componentToTest.find(TaskTitle);
    expect(titleCompo)
    expect(titleCompo.text()).toContain(title)
    expect(titleCompo.text()).not.toContain('what ever text not expected')
  })
  // Enzyme allow us to test the complex implementation related to the React component's instance. Such as the state.
  it('should switch the "completed" state when clicking on the Task', () => {
    // Check the default state of "completed"
    expect(componentToTest.state('completed')).toEqual(false)
    // We might simulate a click event
    componentToTest.find(TaskTitle).simulate('click');
    // Check if the "completed" state is correctly updated
    expect(componentToTest.state('completed')).toEqual(true)
    componentToTest.find(TaskTitle).simulate('click');
    expect(componentToTest.state('completed')).toEqual(false)

  })
})