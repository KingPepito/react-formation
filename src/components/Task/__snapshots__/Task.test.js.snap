// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task snapshot test. Simply test if the component is rendering 1`] = `
<div
  className="sc-AykKD WZuFE"
>
  <span
    className="sc-AykKC kDDRHK"
    data-testid="title-task-test"
    onClick={[Function]}
  >
    Hey I'm a test task!
  </span>
  <button
    data-test="remove-button-task-test"
    onClick={[Function]}
  >
    X
  </button>
</div>
`;
